src/services/index.ts

```diff
@@ -0,0 +1,448 @@
+      const tokenList: string[] = getTokenListFromInputsAndOutputs(txInputs, txOutputs);
+
+      // Update transaction count with the new tx
+      await incrementTokensTxCount(mysql, tokenList);
+
+      const addressBalanceMap: StringMap<TokenBalanceMap> = getAddressBalanceMap(txInputs, txOutputs);
+
+      // update address tables (address, address_balance, address_tx_history)
+      await updateAddressTablesWithTx(mysql, hash, timestamp, addressBalanceMap);
+
+      // for the addresses present on the tx, check if there are any wallets associated
+      const addressWalletMap: StringMap<Wallet> = await getAddressWalletInfo(mysql, Object.keys(addressBalanceMap));
+
+      // for each already started wallet, update databases
+      const seenWallets = new Set();
+      for (const wallet of Object.values(addressWalletMap)) {
+        const walletId = wallet.walletId;
+
+        // this map might contain duplicate wallet values, as 2 different addresses might belong to the same wallet
+        if (seenWallets.has(walletId)) continue;
+        seenWallets.add(walletId);
+        const { newAddresses, lastUsedAddressIndex } = await generateAddresses(mysql, wallet.xpubkey, wallet.maxGap);
+        // might need to generate new addresses to keep maxGap
+        await addNewAddresses(mysql, walletId, newAddresses, lastUsedAddressIndex);
+        // update existing addresses' walletId and index
+      }
+      // update wallet_balance and wallet_tx_history tables
+      const walletBalanceMap: StringMap<TokenBalanceMap> = getWalletBalanceMap(addressWalletMap, addressBalanceMap);
+      await updateWalletTablesWithTx(mysql, hash, timestamp, walletBalanceMap);
+    }
+
+    // TODO: Send message on SQS  for real-time update
+    await dbUpdateLastSyncedEvent(mysql, fullNodeEvent.event.id);
+
+    await mysql.commit();
+  } catch (e) {
+    await mysql.rollback();
+    logger.error(e);
+
+    throw e;
+  } finally {
+    mysql.destroy();
+  }
+};
+
+export const handleVoidedTx = async (context: Context) => {
+  const mysql = await getDbConnection();
+  await mysql.beginTransaction();
+
+  try {
+    const fullNodeEvent = context.event as FullNodeEvent;
+
+    const {
+      hash,
+      outputs,
+      inputs,
+      tokens,
+    } = fullNodeEvent.event.data;
+
+    logger.debug(`Will handle voided tx for ${hash}`);
+
+    const dbTxOutputs: DbTxOutput[] = await getTxOutputsFromTx(mysql, hash);
+    const txOutputs: TxOutputWithIndex[] = prepareOutputs(outputs, tokens);
+    const txInputs: TxInput[] = prepareInputs(inputs, tokens);
+
+    // Set outputs as locked:
+
+    const txOutputsWithLocked = txOutputs.map((output) => {
+      const dbTxOutput = dbTxOutputs.find((_output) => _output.index === output.index);
+
+      if (!dbTxOutput) {
+        throw new Error('Transaction output different from database output!');
+      }
+
+      return {
+        ...output,
+        locked: dbTxOutput.locked,
+      };
+    });
```

Comment:
If we are locking the outputs, shouldn't it be `locked: true` in the returned value?

Reply:
This comment is also misleading, we're not locking the outputs, we're just setting their locked status as it is in the database

Refactored in de59847c68babcbe00bc998df760924ac962ffb0